File: adapters/llm/openai.ts
// adapters/llm/openai.ts — ILLMClient → OpenAI Chat Completions API
import {
  CompletionChunk,
  CompletionRequest,
  CompletionResponse,
  ILLMClient,
  LLMVendor,
  TokenLimitError,
} from '../../core/protocols.js'

export interface OpenAIConfig {
  apiKey: string
  model?: string
  maxTokens?: number
  baseUrl?: string
}

interface OpenAIChoice {
  message: {
    role: string
    content: string | null
    tool_calls?: Array<{
      id: string
      type: string
      function: { name: string; arguments: string }
    }>
  }
  finish_reason: string
}

interface OpenAIChatResponse {
  id: string
  choices: OpenAIChoice[]
  usage: { prompt_tokens: number; completion_tokens: number }
}

export class OpenAILLMClient implements ILLMClient {
  readonly vendor: LLMVendor = 'openai'
  readonly modelId: string
  readonly maxTokens: number
  readonly supportsTools = true

  private readonly _apiKey: string
  private readonly _baseUrl: string

  constructor(config: OpenAIConfig) {
    this._apiKey = config.apiKey
    this.modelId = config.model ?? 'gpt-4o'
    this.maxTokens = config.maxTokens ?? 4096
    this._baseUrl = config.baseUrl ?? 'https://api.openai.com'
  }

  async complete(request: CompletionRequest): Promise<CompletionResponse> {
    if (this._estimateTokens(request) > this.maxTokens) {
      throw new TokenLimitError(this._estimateTokens(request), this.maxTokens)
    }

    const response = await fetch(`${this._baseUrl}/v1/chat/completions`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${this._apiKey}`,
      },
      body: JSON.stringify(this._buildBody(request)),
    })

    if (!response.ok) {
      throw new Error(`OpenAI API error: ${response.status} ${response.statusText}`)
    }

    return this._parse(await response.json() as OpenAIChatResponse)
  }

  async *stream(request: CompletionRequest): AsyncIterableIterator<CompletionChunk> {
    const response = await fetch(`${this._baseUrl}/v1/chat/completions`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${this._apiKey}`,
      },
      body: JSON.stringify({ ...this._buildBody(request), stream: true }),
    })

    if (!response.ok || !response.body) {
      throw new Error(`OpenAI stream error: ${response.status}`)
    }

    const reader = response.body.getReader()
    const decoder = new TextDecoder()

    while (true) {
      const { done, value } = await reader.read()
      if (done) break
      const text = decoder.decode(value)
      for (const line of text.split('\n').filter(l => l.startsWith('data: '))) {
        const json = line.slice(6).trim()
        if (json === '[DONE]') { yield { delta: '', done: true }; return }
        try {
          const event = JSON.parse(json) as {
            choices?: Array<{ delta?: { content?: string } }>
          }
          const delta = event.choices?.[0]?.delta?.content
          if (delta) yield { delta, done: false }
        } catch { /* skip malformed events */ }
      }
    }
    yield { delta: '', done: true }
  }

  private _buildBody(request: CompletionRequest): Record<string, unknown> {
    const body: Record<string, unknown> = {
      model: this.modelId,
      max_tokens: request.maxTokens ?? this.maxTokens,
      messages: request.messages.map(m => ({ role: m.role, content: m.content })),
    }
    if (request.tools?.length) {
      body['tools'] = request.tools
      body['tool_choice'] = 'auto'
    }
    if (request.temperature !== undefined) body['temperature'] = request.temperature
    if (request.stopSequences?.length) body['stop'] = request.stopSequences
    return body
  }

  private _parse(data: OpenAIChatResponse): CompletionResponse {
    const choice = data.choices[0]
    if (!choice) throw new Error('OpenAI returned no choices')

    const toolCalls = choice.message.tool_calls?.map(tc => ({
      id: tc.id,
      name: tc.function.name,
      args: JSON.parse(tc.function.arguments) as Record<string, unknown>,
    }))

    return {
      content: choice.message.content ?? '',
      stopReason: choice.finish_reason === 'tool_calls' ? 'tool_use'
        : choice.finish_reason === 'length' ? 'max_tokens'
        : 'end_turn',
      usage: {
        inputTokens: data.usage.prompt_tokens,
        outputTokens: data.usage.completion_tokens,
      },
      toolCalls,
    }
  }

  private _estimateTokens(request: CompletionRequest): number {
    return Math.ceil(request.messages.reduce((s, m) => s + m.content.length, 0) / 4)
  }
}

File: src/ooagent/adapters/llm/openai.py
"""adapters/llm/openai.py — ILLMClient -> OpenAI Chat Completions API."""

from __future__ import annotations

import codecs
import json
import math
from collections.abc import AsyncIterator
from dataclasses import dataclass
from typing import Any, Literal

import httpx

from ooagent.core.protocols import (
    CompletionChunk,
    CompletionRequest,
    CompletionResponse,
    ILLMClient,
    LLMVendor,
    TokenLimitError,
    TokenUsage,
    ToolCall,
)


@dataclass(frozen=True)
class OpenAIConfig:
    api_key: str
    model: str | None = None
    max_tokens: int | None = None
    base_url: str | None = None


class OpenAILLMClient(ILLMClient):
    """ILLMClient adapter for the OpenAI Chat Completions API."""

    def __init__(self, config: OpenAIConfig) -> None:
        self._api_key = config.api_key
        self._model_id = config.model if config.model is not None else "gpt-4o"
        self._max_tokens = config.max_tokens if config.max_tokens is not None else 4096
        self._base_url = (
            config.base_url if config.base_url is not None else "https://api.openai.com"
        )

    @property
    def vendor(self) -> LLMVendor:
        return "openai"

    @property
    def model_id(self) -> str:
        return self._model_id

    @property
    def max_tokens(self) -> int:
        return self._max_tokens

    @property
    def supports_tools(self) -> bool:
        return True

    async def complete(self, request: CompletionRequest) -> CompletionResponse:
        estimated = self._estimate_tokens(request)
        if estimated > self.max_tokens:
            raise TokenLimitError(estimated, self.max_tokens)

        async with httpx.AsyncClient() as client:
            response = await client.post(
                f"{self._base_url}/v1/chat/completions",
                headers=self._headers(),
                json=self._build_body(request),
            )

        if response.status_code >= 400:
            raise RuntimeError(f"OpenAI API error: {response.status_code} {response.reason_phrase}")

        return self._parse(response.json())

    async def ping(self) -> bool:
        return True

    async def stream(self, request: CompletionRequest) -> AsyncIterator[CompletionChunk]:
        body = {**self._build_body(request), "stream": True}
        async with httpx.AsyncClient() as client:
            async with client.stream(
                "POST",
                f"{self._base_url}/v1/chat/completions",
                headers=self._headers(),
                json=body,
            ) as response:
                if response.status_code >= 400:
                    raise RuntimeError(f"OpenAI stream error: {response.status_code}")

                decoder = codecs.getincrementaldecoder("utf-8")()
                async for raw in response.aiter_bytes():
                    text = decoder.decode(raw)
                    for line in text.split("\n"):
                        if not line.startswith("data: "):
                            continue
                        payload = line[6:].strip()
                        if payload == "[DONE]":
                            yield CompletionChunk(delta="", done=True)
                            return
                        try:
                            event = json.loads(payload)
                        except json.JSONDecodeError:
                            continue  # skip malformed events
                        choices = event.get("choices") or []
                        delta = (choices[0].get("delta") if choices else None) or {}
                        content = delta.get("content")
                        if content:
                            yield CompletionChunk(delta=content, done=False)
        yield CompletionChunk(delta="", done=True)

    def _headers(self) -> dict[str, str]:
        return {
            "Content-Type": "application/json",
            "Authorization": f"Bearer {self._api_key}",
        }

    def _build_body(self, request: CompletionRequest) -> dict[str, Any]:
        body: dict[str, Any] = {
            "model": self.model_id,
            "max_tokens": (
                request.max_tokens if request.max_tokens is not None else self.max_tokens
            ),
            "messages": [{"role": m.role, "content": m.content} for m in request.messages],
        }
        if request.tools:
            body["tools"] = request.tools
            body["tool_choice"] = "auto"
        if request.temperature is not None:
            body["temperature"] = request.temperature
        if request.stop_sequences:
            body["stop"] = request.stop_sequences
        return body

    def _parse(self, data: dict[str, Any]) -> CompletionResponse:
        choices = data.get("choices") or []
        if not choices:
            raise RuntimeError("OpenAI returned no choices")
        choice = choices[0]
        message = choice.get("message", {})

        raw_tool_calls = message.get("tool_calls")
        tool_calls = None
        if raw_tool_calls:
            tool_calls = [
                ToolCall(
                    id=tc["id"],
                    name=tc["function"]["name"],
                    args=json.loads(tc["function"]["arguments"]),
                )
                for tc in raw_tool_calls
            ]

        finish_reason = choice.get("finish_reason")
        stop_reason: Literal["end_turn", "max_tokens", "tool_use", "stop_sequence"]
        if finish_reason == "tool_calls":
            stop_reason = "tool_use"
        elif finish_reason == "length":
            stop_reason = "max_tokens"
        else:
            stop_reason = "end_turn"

        usage = data.get("usage", {})
        return CompletionResponse(
            content=message.get("content") or "",
            stop_reason=stop_reason,
            usage=TokenUsage(
                input_tokens=usage.get("prompt_tokens", 0),
                output_tokens=usage.get("completion_tokens", 0),
            ),
            tool_calls=tool_calls,
        )

    def _estimate_tokens(self, request: CompletionRequest) -> int:
        return math.ceil(sum(len(m.content) for m in request.messages) / 4)
